### PR TITLE
fix(ioredis): fix instrumentation of ESM-imported ioredis

### PR DESCRIPTION
**Pull request by trentm**
- With an ESM import the top-level object is a Module Namespace Object (https://tc39.es/ecma262/#sec-module-namespace-objects) that has no prototype. For compat, Node.js assigns the usual CommonJS module.exports to `<module>.default`, so use that when this is an ESM module.
- Also, TypeScript translates class properties to assignments in the constructor after the super() call. Because super() can call init() and enable() synchronously, it calls back into 'IORedisInstrumentation' before 'traceSendCommand' was defined. Defining it as a method fixes that issue.

Fixes: #1692

## Checklist

- [ ] Decide whether and how to add a test for this.

### CODE REVIEW

**Comment by trentm** [comment]
Note: I still have this in draft, because:

### 1. unit tests

I haven't added any unit tests for this. Adding a unit test for ESM is a little bit of work. I'd appreciate thoughts/suggestions from others. Options:

1. Skip unit tests for this. ;)
2. Have ioredis.test.ts spawn a separate process that runs a small ioredis-using .mjs file with the `node --experimental-loader=...` argument.
3. Add a separate `npm run test:esm` similar to here: https://github.com/open-telemetry/opentelemetry-js/blob/f2fc0d8787a82f33b1ce7e315759ddc61caf4f52/experimental/packages/opentelemetry-instrumentation/package.json#L51-L53


### 2. lint

I was unable to `npm run lint` locally:

```
% npm run lint

> @opentelemetry/instrumentation-ioredis@0.35.1 lint
> eslint . --ext .ts

TypeError: Error while loading rule 'prettier/prettier': context.getPhysicalFilename is not a function
Occurred while linting /Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis/src/index.ts
    at Object.create (/Users/trentm/tm/opentelemetry-js-contrib2/node_modules/eslint-plugin-prettier/eslint-plugin-prettier.js:124:40)
    at createRuleListeners (/Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis/node_modules/eslint/lib/linter/linter.js:746:21)
    at /Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis/node_modules/eslint/lib/linter/linter.js:916:31
    at Array.forEach (<anonymous>)
    at runRules (/Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis/node_modules/eslint/lib/linter/linter.js:861:34)
    at Linter._verifyWithoutProcessors (/Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis/node_modules/eslint/lib/linter/linter.js:1157:31)
    at Linter._verifyWithConfigArray (/Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis/node_modules/eslint/lib/linter/linter.js:1255:21)
    at Linter.verify (/Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis/node_modules/eslint/lib/linter/linter.js:1210:25)
    at Linter.verifyAndFix (/Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis/node_modules/eslint/lib/linter/linter.js:1400:29)
    at verifyText (/Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis/node_modules/eslint/lib/cli-engine/cli-engine.js:230:48)
```

I don't know what that is about.

**Comment by trentm** [comment]
I've resolved lint... by manually pointing to the eslint installed at the top level:

```
$ cd plugins/node/opentelemetry-instrumentation-ioredis
$ ../../../node_modules/.bin/eslint . --ext .ts
```

FWIW, just running `npm run lint` in that dir breaks on, I think, https://github.com/prettier/eslint-plugin-prettier/issues/434 because there is an *old* incompatible eslint installed in this sub-package dir because of a transitive dep of `test-all-versions`:

```
% npm ls eslint
@opentelemetry/instrumentation-ioredis@0.35.1 /Users/trentm/tm/opentelemetry-js-contrib2/plugins/node/opentelemetry-instrumentation-ioredis
└─┬ test-all-versions@5.0.1
  └─┬ parse-env-string@1.0.0
    └─┬ standard@14.3.4
...
      └── eslint@6.8.0
```

**Comment by Flarna** [comment]
regarding the old eslint - this seems to be known issue since a while see [here](https://github.com/watson/parse-env-string/issues/1).

**Comment by trentm** [comment]
> regarding the old eslint - this seems to be known issue since a while see [here](https://github.com/watson/parse-env-string/issues/1).

@Flarna Oh. Thanks for chasing that down!  I hadn't made the connection that this had to have been a "dependencies" entry. I'll see if I can get that fixed upstream.

**Comment by trentm** [comment]
> I'll see if I can get that fixed upstream.

Fixed. parse-env-string@1.0.1 is published. Now if I:

```
$ cd plugins/node/opentelemetry-instrumentation-ioredis
$ npm update
$ npm ls parse-env-string  # verify it is at version 1.0.1
```

then `npm run lint` now works. \o/

**Review by blumamir — APPROVED** [review]
overall LGTM.
I never worked with ESM before, so would appreciate another review from someone with more knowledge (maybe @pichlermarc ?)

Is `module[Symbol.toStringTag] === 'Module'` the canonical way to test for ESM vs CJS?

**Comment by trentm** [comment]
> Is `module[Symbol.toStringTag] === 'Module'` the canonical way to test for ESM vs CJS?

I'm not sure about it being canonical.

I touched briefly on this in an "Aside" section of this long comment: https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1692#issuecomment-1728464734
Because the comment is long, it is easy to not see it.

By ECMAScript spec (https://tc39.es/ecma262/#sec-module-namespace-objects) the Module Namespace Object returned from an ESM import has a well-known symbol `@@toStringTag` value of "Module":

> The initial value of the [@@toStringTag](https://tc39.es/ecma262/#sec-well-known-symbols) property is the String value "Module".

That `@@toStringTag` symbol is represented by [Symbol.toStringTag](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag).  This is what makes `Object.prototype.toString.call()` on a Number or Date include that name:

```
> var d = new Date()
undefined
> Object.prototype.toString.call(d)
'[object Date]'
```

It *is* spoofable:

```
% cat foobar.js
class FooBar {
  get [Symbol.toStringTag]() {
    return 'Module';
  }
}
module.exports = new FooBar();

% node
> var mod = require('./foobar.js')
> mod[Symbol.toStringTag]
'Module'
```

I'm not sure if we need to worry about that kind of spoofing -- accidental or intentional.

--- 

Another way to differentiate would be to pass in a `thisIsESMFromImportInTheMiddle` boolean from where the `hookFn` is setup for RITM vs IITM: https://github.com/open-telemetry/opentelemetry-js/blob/f2fc0d8787a82f33b1ce7e315759ddc61caf4f52/experimental/packages/opentelemetry-instrumentation/src/platform/node/instrumentation.ts#L244-L275  (I'm not suggesting that is the actual variable name. :)

(I'm not positive of the longevity of doing this. If I'm understanding recent changes in Node.js core (https://nodejs.org/en/blog/release/v20.6.0#new-nodemodule-api-register-for-module-customization-hooks-new-initialize-hook) we may eventually just have a single hook on a single Node.js core loader that handles loading both CommonJS and ESM.)

File: plugins/node/opentelemetry-instrumentation-ioredis/src/instrumentation.ts [hunk 1]
```diff
@@ -49,7 +49,11 @@ export class IORedisInstrumentation extends InstrumentationBase<any> {
       new InstrumentationNodeModuleDefinition<any>(
         'ioredis',
         ['>1', '<6'],
-        (moduleExports, moduleVersion?: string) => {
+        (module, moduleVersion?: string) => {
+          const moduleExports =
+            module[Symbol.toStringTag] === 'Module'
+              ? module.default // ESM
+              : module; // CommonJS
```
[the remark this answers is not in the document]
Comment (trentm) [reply plugins/node/opentelemetry-instrumentation-ioredis/src/instrumentation.ts:56]: > Will it be patched?

Yes. That works currently as well. Example script showing that: https://gist.github.com/trentm/22c588cbba28f979b17e5a032ca400c5#file-use-ioredis-default-js

Currently, and with this change, we end up patching properties of the top-level `Redis`, and `.default` is a circular reference to that top-level, so it effectively gets patched as well.



> I wonder if we can alternatively just attempt to patch `default` if it is found on `module` regardless of the import type

Yes, we definitely could here, and I started with that patch.  However, I like that the "detect if this is an ESM-imported module" does not depend on the particular module explicitly having a `.default` circular reference to the top-level export.

Some others do have a `.default` reference (e.g. Fastify), but they won't in general.  E.g. a similar issue for `pg` was recently opened (https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1693) and it does not have a `.default`:

```
> var pg = require('pg')
undefined
> pg.default
undefined
```

As well, I'm working on the Bunyan instrumentation for https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1559 and it'll require special handling of an ESM import and does not have a `.default`.

**Comment by pichlermarc** [comment]
> I'm not sure if we need to worry about that kind of spoofing -- accidental or intentional.

The decision from the SIG meeting: we have come to the conclusion that this being spoofed does not need to be considered at this point in time. We also came to the conclusion that this seems like the best way to test for ESM vs CJS.

**Comment by pichlermarc** [comment]
> ### 1. unit tests
> 
> I haven't added any unit tests for this. Adding a unit test for ESM is a little bit of work. I'd appreciate thoughts/suggestions from others. Options:
> 
>     1. Skip unit tests for this. ;)

:slightly_smiling_face: 

>     2. Have ioredis.test.ts spawn a separate process that runs a small ioredis-using .mjs file with the `node --experimental-loader=...` argument.

Do you mean something like

```
import {IORedisInstrumentation} from "@opentelemetry/instrumentation-ioredis";
new IORedisInstrumentation()
import 'ioredis'
setTimeout(() => 0, 1000)
```

I think this could work as a start. 

>     3. Add a separate `npm run test:esm` similar to here: [open-telemetry/opentelemetry-js@`f2fc0d8`/experimental/packages/opentelemetry-instrumentation/package.json#L51-L53](https://github.com/open-telemetry/opentelemetry-js/blob/f2fc0d8787a82f33b1ce7e315759ddc61caf4f52/experimental/packages/opentelemetry-instrumentation/package.json#L51-L53)

That would also work, yes. I think the tests in @opentelemetry/instrumentation are completely separate ones so they don't test all the same things the CJS tests test do. But I think a simple test like this or the one from 2. would work to get this fix out, then we can think of how to do this in a broader scope in #1731

**Comment by trentm** [comment]
> However, I'd also be okay to skip them for now to get the fix out and then we come up with a general solution in #1731

Thanks for the review! Let's discuss on #1731 and #1735 -- my stab at a utility to do option 2.

File: plugins/node/opentelemetry-instrumentation-ioredis/src/instrumentation.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-ioredis/src/instrumentation.ts:56]: @trentm I only now got back to reading this issue, great research and writing.
I am so confused by this ESM-CJS semantics, will learn step by step 🙏

**Pull request merged**